### PR TITLE
Bug fixes - QA feedback

### PR DESCRIPTION
**Pull request by Scollaco**
# What
Fixed bugs found during our QA process:
1- Change email request being made 2x,
2- Go button enabled even if inputs were not valid,
3- Email allowing emails with invalid formats

# How
The `Go` return type was replaced with the `Done` button. This will make the keyboard dismiss, simplifying the logic, and by tapping the `Save` button will be the only way to make the request.

I also replaced the regex used for the email validation with one on this [article](http://www.cocoawithlove.com/2009/06/verifying-that-string-is-email-address.html).
According to the article, `This regular expression is adapted from a version at regular-expressions.info and is a complete verification of RFC 2822.`. This regex validates a much larger number of email formats.

GIF

### CODE REVIEW

File: Library/ViewModels/ChangeEmailViewModel.swift [hunk 1]
```diff
@@ -257,6 +241,5 @@ ChangeEmailViewModelOutputs {
 }
 
 private func shouldEnableSaveButton(email: String?, newEmail: String, password: String) -> Bool {
```
Comment (justinswart) [inline Library/ViewModels/ChangeEmailViewModel.swift]: Do we allow space-only passwords? Otherwise we might want to trim these too:

```suggestion
private func shouldEnableSaveButton(email: String?, newEmail: String, password: String) -> Bool {
  guard email != newEmail else { return false }

  return ![newEmail, password]
    .map { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    .contains(false)
}
```

File: Library/ViewModels/ChangeEmailViewModel.swift [hunk 2]
```diff
@@ -207,14 +200,6 @@ ChangeEmailViewModelOutputs {
...
-  private let changePasswordProperty = MutableProperty<(String, String)?>(nil)
-  public func submitForm(newEmail: String?, password: String?) {
-
-    if let newEmail = newEmail, let password = password {
-      self.changePasswordProperty.value = (newEmail, password)
-    }
-  }
-
   private let saveButtonTappedProperty = MutableProperty<(String, String)?>(nil)
   public func saveButtonTapped(newEmail: String?, password: String?) {
```
Comment (justinswart) [inline Library/ViewModels/ChangeEmailViewModel.swift]: We shouldn't have any logic in the input function itself, it should be handled via signals.

Also - I don't think these values should be passed in this way, we're setting them on every keypress so we should be able to just have combine those signals and grab the values on our save button press using `takeWhen`.

**Comment by Scollaco** [comment]
Yeah, about the 1Password button, it's beyond the scope of this PR, but it's a good idea to bring this up on the next sprints and implement it on a separate PR. Thanks for your feedback, @justinswart !

File: Kickstarter-iOS/Views/Controllers/ChangeEmailViewController.swift [hunk 3]
```diff
@@ -251,9 +241,17 @@ internal final class ChangeEmailViewController: UIViewController {
   }
 
   private func resetFields(string: String) {
-    _ = [self.passwordTextField, self.newEmailTextField]
-      ||> UITextField.lens.text .~ string
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/ChangeEmailViewController.swift]: Why this change?

Comment (Scollaco) [reply Kickstarter-iOS/Views/Controllers/ChangeEmailViewController.swift]: I just got rid of lenses here, and I remember I saw a warning once saying that, to set texts, it's preferred to use the `%~` operator over `.~`.

Comment (justinswart) [reply Kickstarter-iOS/Views/Controllers/ChangeEmailViewController.swift]: Ah right - so in this case that isn't necessary because the string is not being evaluated for localization. In other places we do this when referring to `Strings.*` to be sure that we're evaluating the closure against any changes to the environment.

File: Library/ViewModels/ChangeEmailViewModelTests.swift [hunk 4]
```diff
@@ -19,14 +19,13 @@ final class ChangeEmailViewModelTests: TestCase {
...
-  private let passwordFieldBecomeFirstResponder = TestObserver<Void, NoError>()
-  private let resetFields = TestObserver<String, NoError>()
-  private let shouldSubmitForm = TestObserver<Void, NoError>()
   private let messageLabelViewHiddenObserver = TestObserver<Bool, NoError>()
+  private let passwordFieldBecomeFirstResponder = TestObserver<Void, NoError>()
   private let passwordText = TestObserver<String, NoError>()
   private let resendVerificationEmailViewIsHiddenObserver = TestObserver<Bool, NoError>()
+  private let resetFields = TestObserver<String, NoError>()
   private let saveButtonIsEnabled = TestObserver<Bool, NoError>()
+  private let textFieldAreEnabled = TestObserver<Bool, NoError>()
```
Comment (justinswart) [inline Library/ViewModels/ChangeEmailViewModelTests.swift]: This property name is missing an `s`

**Comment by Scollaco** [comment]
[Change Email Implementation](https://trello.com/c/gaJRDQz4/869-change-email-implementation)
